### ISSUE TITLE
Unicode column names fail at saving

### ISSUE THREAD

**Issue by p3trus** [astropy/astropy#5218]
With astropy 1.2

A unicode encode error is thrown, when a unicode colname is used
```
from astropy.table import Column, Table

t = Table()
t['Δ'] = Column([1.0, 2.0], unit='m', dtype='float32')
t.write('test.dat', format='ascii.ecsv')
```

This raises a `UnicodeEncodeError: 'charmap' codec can't encode character '\u0394' in position 87: character maps to <undefined>`

**Comment by mirca** [comment]
That does not happen here.
FYI, the written file looks like this: 
```
# %ECSV 0.9
# ---
# datatype:
# - {name: "\u0394", unit: m, datatype: float32}
Δ
1.0
2.0
```

**Comment by mirca** [comment]
@p3trus What's your yaml version?

**Comment by saimn** [comment]
This probably works only with python 3, due to limitations of the python 2 csv module.

**Comment by mirca** [comment]
@saimn It works fine with Python 2 also

**Comment by p3trus** [comment]
yaml version is 3.11

**Comment by mirca** [comment]
@p3trus Could you post the complete log error, please?

**Comment by mirca** [comment]
@p3trus The problem might be with the character encoding that is used in your terminal. Could you check if you're using UTF-8?

**Comment by pllim** [comment]
Also try with the dev version. Just in case this has been fixed since the release.

**Comment by p3trus** [comment]
@mirca The full error message, running the code in a jupyter notebook is
```
---------------------------------------------------------------------------
UnicodeEncodeError                        Traceback (most recent call last)
<ipython-input-1-2b0c3f6b7093> in <module>()
      3 t = Table()
      4 t['Δ'] = Column([1.0, 2.0], unit='m', dtype='float32')
----> 5 t.write('test.dat', format='ascii.ecsv')

C:\Users\p3trus\Anaconda3\lib\site-packages\astropy\table\table.py in write(self, *args, **kwargs)
   2347         passed through to the underlying data reader (e.g. `~astropy.io.ascii.write`).
   2348         """
-> 2349         io_registry.write(self, *args, **kwargs)
   2350 
   2351     def copy(self, copy_data=True):

C:\Users\p3trus\Anaconda3\lib\site-packages\astropy\io\registry.py in write(data, *args, **kwargs)
    401 
    402     writer = get_writer(format, data.__class__)
--> 403     writer(data, *args, **kwargs)
    404 
    405 

C:\Users\p3trus\Anaconda3\lib\site-packages\astropy\io\ascii\connect.py in io_write(format, table, filename, **kwargs)
     41     from .ui import write
     42     format = re.sub(r'^ascii\.', '', format)
---> 43     return write(table, filename, format=format, **kwargs)
     44 
     45 

C:\Users\p3trus\Anaconda3\lib\site-packages\astropy\io\ascii\ui.py in write(table, output, format, Writer, fast_writer, **kwargs)
    694     if not hasattr(output, 'write'):
    695         output = open(output, 'w')
--> 696         output.write(outstr)
    697         output.write(os.linesep)
    698         output.close()

C:\Users\p3trus\Anaconda3\lib\encodings\cp1252.py in encode(self, input, final)
     17 class IncrementalEncoder(codecs.IncrementalEncoder):
     18     def encode(self, input, final=False):
---> 19         return codecs.charmap_encode(input,self.errors,encoding_table)[0]
     20 
     21 class IncrementalDecoder(codecs.IncrementalDecoder):

UnicodeEncodeError: 'charmap' codec can't encode character '\u0394' in position 87: character maps to <undefined>

```
@pllim  ​I will try the dev version soon.

**Comment by mirca** [comment]
@p3trus I'm pretty sure that if you change your text encoding to Unicode UTF-8 you will be fine.

**Comment by p3trus** [comment]
How do I change the text encoding in a jupyter notebook?

**Comment by mirca** [comment]
@p3trus you need to change it in your system terminal. What OS do you use?
In OSX, go to Terminal -> Preferences -> Profile -> Text Encoding.
In Windows, see the answer in: http://stackoverflow.com/questions/32382686/unicodeencodeerror-charmap-codec-cant-encode-character-u2010-character-m

**Comment by p3trus** [comment]
@mirca This changes the encoding of a terminal. I'm using a notebook, not sure if that should change anything.

**Comment by mirca** [comment]
@p3trus You trying to write a file to your system, but the encoding you are using does not have a mapping for the '\u0394' character. So, changing the system text encoding should to an appropriate one should matter

**Comment by p3trus** [comment]
@mirca You're right. After reading this https://mail.python.org/pipermail/python-dev/2011-June/112086.html it makes sense.

> In Python 2, open() opens the file in binary mode (e.g. file.readline()
> returns a byte string). codecs.open() opens the file in binary mode by
> default, you have to specify an encoding name to open it in text mode.
> 
> In Python 3, open() opens the file in text mode by default. (It only
> opens the binary mode if the file mode contains "b".) The problem is
> that open() uses the locale encoding if the encoding is not specified,
> which is the case *by default*. The locale encoding can be:
> 
>  - UTF-8 on Mac OS X, most Linux distributions
>  - ISO-8859-1 os some FreeBSD systems
>  - ANSI code page on Windows, e.g. cp1252 (close to ISO-8859-1) in
> Western Europe, cp952 in Japan, ...

Adding an explicit `encoding='utf-8'` to the `open()` call in `astropy.io.ascii.ui.py` works.

```
if not hasattr(output, 'write'):
    output = open(output, 'w', encoding='utf-8')
    output.write(outstr)
    output.write(os.linesep)
    output.close()
```

Adding 
```
import win32console
win32console.SetConsoleCP(65001)
```
works as well, but I don't think that changing the default encoding is a good idea just to be able to write a file.

As a side note, is there a reason the file is not opened as a context manager or wrapped in a try-except block?

**Comment by p3trus** [comment]
@mirca By the way, the same problem occurs on loading a file with unicode symbols. What helps is setting the default locale on python 3.5 like this

    import _locale
    _locale._getdefaultlocale = (lambda *args: ['de_DE', 'utf8'])

But thats a nasty hack. So please add the explicit encoding to the open method calls.

**Comment by p3trus** [comment]
Is there any progress? My code example is still failing with astropy 1.3 on windows 7 and python 3.6 (anaconda)

**Comment by bsipocz** [comment]
@p3trus - Using the dev version your code example (from the opening comment) runs without raising an error for me.

**Comment by bsipocz** [comment]
But just as I see above, it worked for @mirca as well (I got the same output).

**Comment by bsipocz** [comment]
So I'm pinging @taldcroft and @hamogu to identify whether there is an action item here for us.

**Comment by hamogu** [comment]
I leave that for @taldcroft to decide. There is a reason the format is called "ascii.ecsv", "ascii.csv" etc. and not "unicode.ecsv" - it was made for ascii tables and never intended to be used with unicode. The fact tat is works in some cases is great and @taldcroft has invested significant amount of work to drive this forward. 
Personally, I'd leave as is for now in general (works in some cases but not in others depending on system encoding, character table etc.) and really tackle after astropy 2.0 when we drop python 2 support altogether. Making all unicode work in all cases in both python 2 and 3 is an extra effort that we can avoid.

That does not mean that we should not fix small things were we can, just that I would not make this a major item for the next release, but for the release after that.

**Comment by p3trus** [comment]
As i wrote earlier, specifying the encoding in `astropy.io.ascii.ui.py` fixes this. Also windows adds an additional newline character to each line, but thats another bug, (fixed by adding `newline=''` to the `open()` call).

**Comment by saimn** [comment]
@p3trus - What's the default encoding on your system ? (`import sys ; print(sys.getdefaultencoding())`)

On Python 3, `open` should already use utf8 by default on most systems, but I don't know much about Windows ... And for Python 2 it would be better to use `io.open` instead of `open`, so the solution (Py2&3) would be `output = io.open(output, 'w', encoding='utf-8')`.

**Comment by pllim** [comment]
No longer an issue, it seems. I can write out the file in original post with astropy `7.2.dev`

```
# %ECSV 1.0
# ---
# datatype:
# - {name: "\u0394", unit: m, datatype: float32}
# schema: astropy-2.0
Δ
1.0
2.0
```
